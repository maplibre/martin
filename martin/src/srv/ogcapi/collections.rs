//! OGC API collections endpoints

use actix_web::http::header::CONTENT_TYPE;
use actix_web::web::{Data, Path};
use actix_web::{HttpRequest, HttpResponse, Result as ActixResult, route};
use log::warn;
use martin_core::tiles::Source;
use ogcapi_types::common::{
    Bbox, Collection as OgcCollection, Collections as OgcCollections, Crs, Extent as OgcExtent,
    Link, SpatialExtent as OgcSpatialExtent,
};
use ogcapi_types::tiles::{
    BoundingBox2D, DataType, GeospatialData, TileMatrixLimits, TilePoint, TileSet, TileSetItem,
    TileSets, TitleDescriptionKeywords,
};
use serde::Deserialize;
use serde_json::Map;

use super::landing::get_base_url;
use crate::source::TileSources;
use crate::srv::server::Catalog;

#[derive(Deserialize)]
pub struct CollectionPath {
    collection_id: String,
}

#[derive(Deserialize)]
pub struct CollectionTileMatrixSetPath {
    collection_id: String,
    tilematrixset_id: String,
}

/// Convert tilejson bounds to OGC API spatial extent
fn bounds_to_spatial_extent(bounds: &tilejson::Bounds) -> OgcSpatialExtent {
    // Create a Bbox2D with the bounds in [west, south, east, north] order
    let bbox = Bbox::Bbox2D([bounds.left, bounds.bottom, bounds.right, bounds.top]);

    // Construct the spatial extent directly
    OgcSpatialExtent {
        bbox: vec![bbox],
        crs: Crs::new(ogcapi_types::common::Authority::OGC, "1.3", "CRS84"),
    }
}

/// Create an OGC API Collection from a tile source
fn create_collection(
    id: String,
    source_result: Result<Box<dyn Source>, actix_web::Error>,
    base_url: &str,
) -> Option<OgcCollection> {
    match source_result {
        Ok(source) => {
            let tj = source.get_tilejson();

            // Build the collection using ogcapi_types::common::Collection
            let collection = OgcCollection {
                id: id.clone(),
                title: tj.name.clone(),
                description: tj.description.clone(),
                keywords: vec![],
                attribution: tj.attribution.clone(),
                extent: tj.bounds.map(|bounds| OgcExtent {
                    spatial: Some(bounds_to_spatial_extent(&bounds)),
                    temporal: None,
                }),
                item_type: Some("tile".to_string()),
                crs: vec![Crs::from_epsg(3857)],
                storage_crs: Some(Crs::from_epsg(3857)),
                storage_crs_coordinate_epoch: None,
                links: vec![
                    Link::new(format!("{base_url}/ogc/collections/{id}"), "self")
                        .mediatype("application/json")
                        .title(format!("Collection {id}")),
                    Link::new(
                        format!("{base_url}/ogc/collections/{id}/tiles"),
                        "http://www.opengis.net/def/rel/ogc/1.0/tilesets-vector",
                    )
                    .mediatype("application/json")
                    .title(format!("Vector tilesets for collection {id}")),
                ],
                additional_properties: {
                    let mut props = Map::new();
                    props.insert("dataType".to_string(), serde_json::json!(DataType::Vector));
                    props
                },
            };

            Some(collection)
        }
        Err(e) => {
            warn!("Failed to get source for collection '{id}': {e}");
            None
        }
    }
}

/// Create OGC API Collections response from catalog and sources
fn create_collections(_catalog: &Catalog, sources: &TileSources, base_url: &str) -> OgcCollections {
    let mut collection = OgcCollections::new(
        sources
            .source_names()
            .into_iter()
            .filter_map(|id| {
                let source_result = sources.get_source(&id);
                create_collection(id, source_result, base_url)
            })
            .collect(),
    );
    collection.links = vec![
        Link::new(format!("{base_url}/ogc/collections"), "self")
            .mediatype("application/json")
            .title("All Collections"),
    ];
    collection
}

/// OGC API Collections endpoint
#[route("/ogc/collections", method = "GET", method = "HEAD")]
pub async fn get_collections(
    req: HttpRequest,
    _catalog: Data<Catalog>,
    sources: Data<TileSources>,
) -> ActixResult<HttpResponse> {
    let base_url = get_base_url(&req);
    let collections = create_collections(&_catalog, &sources, &base_url);

    Ok(HttpResponse::Ok()
        .insert_header((CONTENT_TYPE, "application/json"))
        .json(collections))
}

/// OGC API Collection endpoint
#[route("/ogc/collections/{collection_id}", method = "GET", method = "HEAD")]
pub async fn get_collection(
    req: HttpRequest,
    path: Path<CollectionPath>,
    _catalog: Data<Catalog>,
    sources: Data<TileSources>,
) -> ActixResult<HttpResponse> {
    let base_url = get_base_url(&req);

    // Get the specific collection
    let collection = create_collection(
        path.collection_id.clone(),
        sources.get_source(&path.collection_id),
        &base_url,
    )
    .ok_or_else(|| actix_web::error::ErrorNotFound("Collection not found"))?;

    Ok(HttpResponse::Ok()
        .insert_header((CONTENT_TYPE, "application/json"))
        .json(collection))
}

/// OGC API Collection Tiles endpoint (list of tilesets for a collection)
#[route(
    "/ogc/collections/{collection_id}/tiles",
    method = "GET",
    method = "HEAD"
)]
pub async fn get_collection_tiles(
    req: HttpRequest,
    path: Path<CollectionPath>,
    sources: Data<TileSources>,
) -> ActixResult<HttpResponse> {
    let base_url = get_base_url(&req);

    // Check if collection exists
    let source = sources
        .get_source(&path.collection_id)
        .map_err(|_| actix_web::error::ErrorNotFound("Collection not found"))?;

    let tj = source.get_tilejson();

    // Create a tileset item for this collection
    let tileset = TileSetItem {
        title: tj.name.clone(),
        data_type: DataType::Vector,
        crs: Crs::from_epsg(3857),
        tile_matrix_set_uri: Some(
            "http://www.opengis.net/def/tilematrixset/OGC/1.0/WebMercatorQuad".to_string(),
        ),
        links: vec![
            Link {
                rel: "self".to_string(),
                r#type: Some("application/json".to_string()),
                title: Some(format!("Tileset for {}", path.collection_id)),
                href: format!(
                    "{}/ogc/collections/{}/tiles/WebMercatorQuad",
                    base_url, path.collection_id
                ),
                hreflang: None,
                length: None,
            },
            Link {
                rel: "http://www.opengis.net/def/rel/ogc/1.0/tiling-scheme".to_string(),
                r#type: Some("application/json".to_string()),
                title: Some("TileMatrixSet".to_string()),
                href: format!("{}/ogc/tileMatrixSets/WebMercatorQuad", base_url),
                hreflang: None,
                length: None,
            },
            Link {
                rel: "item".to_string(),
                r#type: Some("application/vnd.mapbox-vector-tile".to_string()),
                title: Some("Tiles".to_string()),
                href: format!(
                    "{}/ogc/collections/{}/tiles/WebMercatorQuad/{{tileMatrix}}/{{tileRow}}/{{tileCol}}",
                    base_url, path.collection_id
                ),
                hreflang: None,
                length: None,
            },
        ],
    };

    let response = TileSets {
        tilesets: vec![tileset],
        links: Some(vec![Link {
            rel: "self".to_string(),
            r#type: Some("application/json".to_string()),
            title: Some("This document".to_string()),
            href: format!("{}/ogc/collections/{}/tiles", base_url, path.collection_id),
            hreflang: None,
            length: None,
        }]),
    };

    Ok(HttpResponse::Ok()
        .insert_header((CONTENT_TYPE, "application/json"))
        .json(response))
}

/// OGC API Collection Tileset endpoint
#[route(
    "/ogc/collections/{collection_id}/tiles/{tilematrixset_id}",
    method = "GET",
    method = "HEAD"
)]
pub async fn get_collection_tileset(
    req: HttpRequest,
    path: Path<CollectionTileMatrixSetPath>,
    sources: Data<TileSources>,
) -> ActixResult<HttpResponse> {
    let base_url = get_base_url(&req);

    // For now, we only support WebMercatorQuad
    if path.tilematrixset_id != "WebMercatorQuad" {
        return Err(actix_web::error::ErrorNotFound("TileMatrixSet not found"));
    }

    // Check if collection exists
    let source = sources
        .get_source(&path.collection_id)
        .map_err(|_| actix_web::error::ErrorNotFound("Collection not found"))?;

    let tj = source.get_tilejson();

    let tileset = TileSet {
        title_description_keywords: TitleDescriptionKeywords {
            title: tj.name.clone(),
            description: tj.description.clone(),
            keywords: None,
        },
        data_type: DataType::Vector,
        tile_matrix_set_uri: Some(
            "http://www.opengis.net/def/tilematrixset/OGC/1.0/WebMercatorQuad".to_string(),
        ),
        tile_matrix_set_limits: tj.minzoom.and_then(|min| {
            tj.maxzoom.map(|max| {
                (min..=max)
                    .map(|z| TileMatrixLimits {
                        tile_matrix: z.to_string(),
                        min_tile_row: 0,
                        max_tile_row: ((1 << z) - 1) as u64,
                        min_tile_col: 0,
                        max_tile_col: ((1 << z) - 1) as u64,
                    })
                    .collect()
            })
        }),
        crs: Crs::from_epsg(3857),
        epoch: None,
        links: vec![
            Link {
                rel: "self".to_string(),
                r#type: Some("application/json".to_string()),
                title: Some("This document".to_string()),
                href: format!(
                    "{}/ogc/collections/{}/tiles/{}",
                    base_url, path.collection_id, path.tilematrixset_id
                ),
                hreflang: None,
                length: None,
            },
            Link {
                rel: "http://www.opengis.net/def/rel/ogc/1.0/tiling-scheme".to_string(),
                r#type: Some("application/json".to_string()),
                title: Some("TileMatrixSet".to_string()),
                href: format!("{}/ogc/tileMatrixSets/{}", base_url, path.tilematrixset_id),
                hreflang: None,
                length: None,
            },
            Link {
                rel: "item".to_string(),
                r#type: Some("application/vnd.mapbox-vector-tile".to_string()),
                title: Some("Mapbox Vector Tiles".to_string()),
                href: format!(
                    "{}/ogc/collections/{}/tiles/{}/{{tileMatrix}}/{{tileRow}}/{{tileCol}}",
                    base_url, path.collection_id, path.tilematrixset_id
                ),
                hreflang: None,
                length: None,
            },
        ],
        layers: tj.vector_layers.clone().map(|layers| {
            layers
                .into_iter()
                .map(|layer| GeospatialData {
                    title_description_keywords: TitleDescriptionKeywords {
                        title: None,
                        description: layer.description,
                        keywords: None,
                    },
                    id: layer.id,
                    data_type: DataType::Vector,
                    geometry_dimension: None,
                    feature_type: None,
                    point_of_contact: None,
                    publisher: None,
                    theme: None,
                    crs: None,
                    epoch: None,
                    min_scale_denominator: None,
                    max_scale_denominator: None,
                    min_cell_size: None,
                    max_cell_size: None,
                    max_tile_matrix: None,
                    min_tile_matrix: None,
                    bounding_box: None,
                    created: None,
                    updated: None,
                    style: None,
                    geo_data_classes: None,
                    properties_schema: None,
                    links: None,
                })
                .collect()
        }),
        bounding_box: tj.bounds.map(|b| BoundingBox2D {
            lower_left: [b.left, b.bottom],
            upper_right: [b.right, b.top],
            crs: None,
            ordered_axes: None,
        }),
        style: None,
        center_point: tj.center.map(|c| TilePoint {
            coordinates: Some([c.longitude, c.latitude]),
            crs: None,
            tile_matrix: Some(c.zoom.to_string()),
            scale_denominator: None,
            cell_size: None,
        }),
        license: None,
        access_constraints: None,
        version: None,
        created: None,
        updated: None,
        point_of_contact: None,
        media_types: Some(vec!["application/vnd.mapbox-vector-tile".to_string()]),
    };

    Ok(HttpResponse::Ok()
        .insert_header((CONTENT_TYPE, "application/json"))
        .json(tileset))
}
