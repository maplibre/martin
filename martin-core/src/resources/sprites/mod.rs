//! Sprite processing and serving for map tile rendering.
//!
//! Generates spritesheets from SVG files with support for high-DPI (@2x) and
//! SDF (Signed Distance Field) sprites for dynamic styling.
//!
//! # Usage
//!
//! ```rust,no_run
//! # async fn foo() {
//! use martin_core::sprites::SpriteSources;
//! use std::path::PathBuf;
//!
//! let mut sources = SpriteSources::default();
//! sources.add_source("icons".to_string(), PathBuf::from("/path/to/svg/directory"));
//! let spritesheet = sources.get_sprites("icons@2x", false).await.unwrap();
//! # }
//! ```

use std::collections::HashMap;
use std::fmt::Debug;
use std::path::PathBuf;

use dashmap::{DashMap, Entry};
use futures::future::try_join_all;
use log::{info, warn};
use serde::{Deserialize, Serialize};
pub use spreet::Spritesheet;
use spreet::resvg::usvg::{Options, Tree};
use spreet::{Sprite, SpritesheetBuilder, get_svg_input_paths, sprite_name};
use tokio::io::AsyncReadExt;

use self::SpriteError::{SpriteInstError, SpriteParsingError, SpriteProcessingError};

mod error;
pub use error::SpriteError;

mod cache;
pub use cache::{NO_SPRITE_CACHE, OptSpriteCache, SpriteCache};

/// Sprite source metadata.
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct CatalogSpriteEntry {
    /// Available sprite image names.
    pub images: Vec<String>,
}

/// Catalog mapping sprite names to metadata (e.g., "icons" -> [`CatalogSpriteEntry`]).
pub type SpriteCatalog = HashMap<String, CatalogSpriteEntry>;

/// Thread-safe sprite source manager for serving sprites as `.png` or `.json`.
#[derive(Debug, Clone, Default)]
pub struct SpriteSources(DashMap<String, SpriteSource>);

impl SpriteSources {
    /// Returns a catalog of all sprite sources.
    pub fn get_catalog(&self) -> Result<SpriteCatalog, SpriteError> {
        // TODO: all sprite generation should be pre-cached
        let mut entries = SpriteCatalog::new();
        for source in &self.0 {
            let paths = get_svg_input_paths(&source.path, true)
                .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
            let mut images = Vec::with_capacity(paths.len());
            for path in paths {
                images.push(
                    sprite_name(&path, &source.path)
                        .map_err(|e| SpriteProcessingError(e, source.path.clone()))?,
                );
            }
            images.sort();
            entries.insert(source.key().clone(), CatalogSpriteEntry { images });
        }
        Ok(entries)
    }

    /// Adds a sprite source directory containing SVG files.
    /// Files are ignored - only directories accepted. Duplicates ignored with warning.
    pub fn add_source(&mut self, id: String, path: PathBuf) {
        let disp_path = path.display();
        if path.is_file() {
            warn!("Ignoring non-directory sprite source {id} from {disp_path}");
        } else {
            match self.0.entry(id) {
                Entry::Occupied(v) => {
                    warn!(
                        "Ignoring duplicate sprite source {} from {disp_path} because it was already configured for {}",
                        v.key(),
                        v.get().path.display()
                    );
                }
                Entry::Vacant(v) => {
                    info!("Configured sprite source {} from {disp_path}", v.key());
                    v.insert(SpriteSource { path });
                }
            }
        }
    }

    /// Generates a spritesheet from comma-separated sprite source IDs.
    ///
    /// Append "@2x" for high-DPI sprites.
    /// Set `as_sdf` for SDF sprites.
    pub async fn get_sprites(&self, ids: &str, as_sdf: bool) -> Result<Spritesheet, SpriteError> {
        let (ids, dpi) = if let Some(ids) = ids.strip_suffix("@2x") {
            (ids, 2)
        } else {
            (ids, 1)
        };

        let sprite_ids = ids
            .split(',')
            .map(|id| self.get(id))
            .collect::<Result<Vec<_>, SpriteError>>()?;

        get_spritesheet(sprite_ids.iter(), dpi, as_sdf).await
    }

    fn get(&self, id: &str) -> Result<SpriteSource, SpriteError> {
        match self.0.get(id) {
            Some(v) => Ok(v.clone()),
            None => Err(SpriteError::SpriteNotFound(id.to_string())),
        }
    }
}

/// Sprite source directory.
#[derive(Clone, Debug)]
pub struct SpriteSource {
    path: PathBuf,
}

/// Parses SVG file into sprite.
async fn parse_sprite(
    name: String,
    path: PathBuf,
    pixel_ratio: u8,
    as_sdf: bool,
) -> Result<(String, Sprite), SpriteError> {
    let on_err = |e| SpriteError::IoError(e, path.clone());

    let mut file = tokio::fs::File::open(&path).await.map_err(on_err)?;

    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).await.map_err(on_err)?;

    let tree = Tree::from_data(&buffer, &Options::default())
        .map_err(|e| SpriteParsingError(e, path.clone()))?;

    let sprite = if as_sdf {
        Sprite::new_sdf(tree, pixel_ratio)
    } else {
        Sprite::new(tree, pixel_ratio)
    };
    let sprite = sprite.ok_or_else(|| SpriteInstError(path.clone()))?;

    Ok((name, sprite))
}

/// Generates spritesheet from sprite sources.
pub async fn get_spritesheet(
    sources: impl Iterator<Item = &SpriteSource>,
    pixel_ratio: u8,
    as_sdf: bool,
) -> Result<Spritesheet, SpriteError> {
    // Asynchronously load all SVG files from the given sources
    let mut futures = Vec::new();
    for source in sources {
        let paths = get_svg_input_paths(&source.path, true)
            .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
        for path in paths {
            let name = sprite_name(&path, &source.path)
                .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
            futures.push(parse_sprite(name, path, pixel_ratio, as_sdf));
        }
    }
    let sprites = try_join_all(futures).await?;
    let mut builder = SpritesheetBuilder::new();
    if as_sdf {
        builder.make_sdf();
    }
    builder.sprites(sprites.into_iter().collect());

    // TODO: decide if this is needed and/or configurable
    // builder.make_unique();

    builder
        .generate()
        .ok_or(SpriteError::UnableToGenerateSpritesheet)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_sprites() {
        let mut sprites = SpriteSources::default();
        sprites.add_source(
            "src1".to_string(),
            PathBuf::from("../tests/fixtures/sprites/src1"),
        );
        sprites.add_source(
            "src2".to_string(),
            PathBuf::from("../tests/fixtures/sprites/src2"),
        );

        assert_eq!(sprites.0.len(), 2);

        for generate_sdf in [true, false] {
            let paths = sprites
                .0
                .iter()
                .map(|v| v.value().clone())
                .collect::<Vec<_>>();
            test_src(paths.iter(), 1, "all_1", generate_sdf).await;
            test_src(paths.iter(), 2, "all_2", generate_sdf).await;

            let src1_path = sprites.get("src1").into_iter().collect::<Vec<_>>();
            test_src(src1_path.iter(), 1, "src1_1", generate_sdf).await;
            test_src(src1_path.iter(), 2, "src1_2", generate_sdf).await;

            let src2_path = sprites.get("src2").into_iter().collect::<Vec<_>>();
            test_src(src2_path.iter(), 1, "src2_1", generate_sdf).await;
            test_src(src2_path.iter(), 2, "src2_2", generate_sdf).await;
        }
    }

    async fn test_src(
        sources: impl Iterator<Item = &SpriteSource>,
        pixel_ratio: u8,
        filename: &str,
        generate_sdf: bool,
    ) {
        let sprites = get_spritesheet(sources, pixel_ratio, generate_sdf)
            .await
            .unwrap();
        let filename = if generate_sdf {
            format!("{filename}_sdf")
        } else {
            filename.to_string()
        };
        insta::assert_json_snapshot!(format!("{filename}.json"), sprites.get_index());
        let png = sprites.encode_png().unwrap();
        insta::assert_binary_snapshot!(&format!("{filename}.png"), png);
    }
}
