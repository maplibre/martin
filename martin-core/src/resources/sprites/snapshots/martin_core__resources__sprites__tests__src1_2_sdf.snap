---
source: martin-core/src/resources/sprites/mod.rs
expression: png
extension: png
snapshot_kind: binary
---
